### PR TITLE
feat(sql): Implement ST_StartPoint() and ST_EndPoint()

### PR DESCRIPTION
**Pull request by yutannihilation**
As many people are adding geos-related functions, I pick this to avoid collision.

~~At first, I thought this can be implemented via `geo_traits::CoordTrait`. But, I read the implementation of `ST_Point` and felt it's preferable to handle the raw bytes of WKB for performance. While I'm not good at performance, I'm trying to implement it.~~

The remaining tasks are:

- [x] Add Python tests
- [x] Add benchmark

### CODE REVIEW

**Comment by yutannihilation** [comment]
Thanks for the hint! Now I figured out how to do it with geo-traits and the functions in wkb_factory.rs.

**Comment by yutannihilation** [comment]
It seems I still have troubles on running Python tests on my Windows, so I don't run these Python tests yet. I'll try on macOS tomorrow. Anyway, I believe this pull request is ready for review now.

File: python/sedonadb/tests/functions/test_functions.py [hunk 1]
```diff
@@ -1015,6 +1015,45 @@ def test_st_pointm(eng, x, y, m, expected):
...
+@pytest.mark.parametrize(
+    ("geometry", "expected"),
+    [
+        ("LINESTRING (1 2, 3 4, 5 6)",                "POINT (1 2)"       ),
+        ("LINESTRING Z (1 2 3, 3 4 5, 5 6 7)",        "POINT Z (1 2 3)"   ),
+        ("LINESTRING M (1 2 3, 3 4 5, 5 6 7)",        "POINT M (1 2 3)"   ),
+        ("LINESTRING ZM (1 2 3 4, 3 4 5 6, 5 6 7 8)", "POINT ZM (1 2 3 4)"),
+        ("POINT (1 2)",                               None                ),
+        ("POLYGON ((0 0, 10 0, 10 10, 0 10, 0 0))",   None                ),
+        ("MULTILINESTRING ((1 2, 3 4), (5 6, 7 8))",  None                ),                ),
```
Comment (paleolimbot) [inline python/sedonadb/tests/functions/test_functions.py]:
```suggestion
        ("MULTILINESTRING ((1 2, 3 4), (5 6, 7 8))",  None                ),
```

File: rust/sedona-functions/src/st_start_point.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+                        line_string.coord(0)
+                    } else {
+                        line_string.coord(line_string.num_coords() - 1)
+                    };
+
+                    if let Some(coord) = maybe_coord {
+                        write_wkb_start_point(&mut builder, coord).map_err(|_| {
+                            datafusion_common::DataFusionError::Internal(
+                                "Failed to write WKB point header".to_string(),
+                            )
```
Comment (paleolimbot) [inline rust/sedona-functions/src/st_start_point.rs]:
```suggestion
                            sedona_internal_err!("Failed to write WKB point header")

```

File: rust/sedona-functions/src/st_start_point.rs [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                coords.y(),
+                coords.nth_or_panic(2),
+                coords.nth_or_panic(3),
+            );
+            write_wkb_coord(buf, coords_tuple)
+        }
+        _ => Err(SedonaGeometryError::Invalid(
+            "Unsupported number of dimensions".to_string(),
+        )),
+    }
```
Comment (paleolimbot) [inline rust/sedona-functions/src/st_start_point.rs]: This would fit nicely as `wkb_factory::write_wkb_coord_trait()` (I'm sure we'll need it again!)

**Comment by yutannihilation** [comment]
Thanks for reviewing!

Sorry, this was unexpected, but I just found [the document](https://postgis.net/docs/en/ST_StartPoint.html) has this small note

> Enhanced: 3.2.0 returns a point for all geometries. Prior behavior returns NULLs if input was not a LineString.

and this is the actual behavior (while the example on the document returns `t`...)

```
postgres=# SELECT ST_StartPoint('POINT(0 1)'::geometry) IS NULL AS is_null;
 is_null 
---------
 f
(1 row)
```

Probably I need some time to update the implementation.

File: rust/sedona-functions/src/st_start_point.rs (hunk 3 above)
Comment (yutannihilation) [reply rust/sedona-functions/src/st_start_point.rs]: Thanks, moved the implementation to wkb_factory now.

**Comment by yutannihilation** [comment]
I think this should be ready for review again now!

Benchmark result:

```
--------------------------------------------------------------------------------------------------------- benchmark: 6 tests --------------------------------------------------------------------------------------------------------
Name (time in ms)                                                     Min                 Max                Mean            StdDev              Median               IQR            Outliers       OPS            Rounds  Iterations
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_st_start_point[segments_large-SedonaDBSingleThread]           1.9418 (1.0)        2.3150 (1.0)        2.0126 (1.0)      0.0812 (1.0)        1.9881 (1.0)      0.0475 (1.0)           2;2  496.8624 (1.0)          26           1
test_st_start_point[collections_simple-DuckDBSingleThread]         6.3318 (3.26)       7.5675 (3.27)       6.7992 (3.38)     0.4783 (5.89)       6.6988 (3.37)     0.6017 (12.66)         1;0  147.0758 (0.30)          5           1
test_st_start_point[segments_large-DuckDBSingleThread]             7.6638 (3.95)       8.6637 (3.74)       7.9898 (3.97)     0.3350 (4.12)       7.8963 (3.97)     0.2996 (6.30)          1;1  125.1591 (0.25)          7           1
test_st_start_point[collections_simple-SedonaDBSingleThread]      10.4538 (5.38)      11.3186 (4.89)      10.6016 (5.27)     0.2383 (2.93)      10.5149 (5.29)     0.0893 (1.88)          2;3   94.3255 (0.19)         20           1
test_st_start_point[segments_large-PostGISSingleThread]          240.1957 (123.70)   243.3285 (105.11)   241.3920 (119.94)   1.1893 (14.64)    241.1900 (121.32)   1.3549 (28.50)         2;0    4.1426 (0.01)          5           1
test_st_start_point[collections_simple-PostGISSingleThread]      352.2113 (181.39)   357.8562 (154.58)   355.9990 (176.88)   2.3396 (28.80)    357.1065 (179.63)   3.0995 (65.20)         1;0    2.8090 (0.01)          5           1
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

File: rust/sedona-functions/src/st_start_point.rs [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        vec![Arc::new(STStartOrEndPoint::new(true))],
+        Volatility::Immutable,
+        Some(st_start_point_doc()),
+    )
+}
+
+fn st_start_point_doc() -> Documentation {
+    Documentation::builder(
+        DOC_SECTION_OTHER,
+        "Returns the start point of a LINESTRING geometry. Returns NULL if the geometry is not a LINESTRING.",
```
Comment (martin-g) [inline rust/sedona-functions/src/st_start_point.rs]: According to https://github.com/apache/sedona-db/pull/245/files#diff-6675d3a4b31524e6b409c5e557bdecd20adb0615fba805d77ba4b52a61c4dae5R142-R143 this is correct for the end point, not for the start one.

File: python/sedonadb/tests/functions/test_functions.py [hunk 5]
```diff
@@ -1016,6 +1016,55 @@ def test_st_pointm(eng, x, y, m, expected):
     )
 
 
+@pytest.mark.parametrize("eng", [SedonaDB, PostGIS])
+@pytest.mark.parametrize(
+    ("geometry", "expected"),
+    [
+        ("LINESTRING (1 2, 3 4, 5 6)", "POINT (1 2)"),
```
Comment (paleolimbot) [inline python/sedonadb/tests/functions/test_functions.py:1031]: I think it would be good to verify that we have the same behaviour as PostGIS for empties as well:

```suggestion
        (None, None),
        ("POINT EMPTY", None),
        ("LINESTRING EMPTY", None),
        ("POLYGON EMPTY", None),
        ("MULTIPOINT EMPTY", None),
        ("MULTILINESTRING EMPTY", None),
        ("MULTIPOLYGON EMPTY", None),
        ("GEOMETRYCOLLECTION EMPTY", None),
        ("LINESTRING (1 2, 3 4, 5 6)", "POINT (1 2)"),
```

File: python/sedonadb/tests/functions/test_functions.py [hunk 6]
```diff
@@ -1016,6 +1016,55 @@ def test_st_pointm(eng, x, y, m, expected):
...
+        f"SELECT ST_StartPoint({geom_or_null(geometry)})",
+        expected,
+    )
+
+
+@pytest.mark.parametrize("eng", [SedonaDB, PostGIS])
+@pytest.mark.parametrize(
+    ("geometry", "expected"),
+    [
+        ("LINESTRING (1 2, 3 4, 5 6)", "POINT (5 6)"),
```
Comment (paleolimbot) [inline python/sedonadb/tests/functions/test_functions.py:1067]: NULL and empties would be good to test here, too.

File: rust/sedona-functions/src/st_start_point.rs [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+    buf: &mut impl Write,
+    coord: impl CoordTrait<T = f64>,
+) -> Result<(), SedonaGeometryError> {
+    write_wkb_point_header(buf, coord.dim())?;
+    write_wkb_coord_trait(buf, &coord)
+}
+
+// - ST_StartPoint returns result for all types of geometries
+// - ST_EndPoint returns result only for LINESTRING
+fn extract_first_geometry<'a>(
```
Comment (paleolimbot) [inline rust/sedona-functions/src/st_start_point.rs]: Optional nit:

```suggestion
fn extract_start_or_end_coord<'a>(

```

File: rust/sedona-functions/src/st_start_point.rs [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+
+            builder.append_null();
+            Ok(())
+        })?;
+
+        executor.finish(Arc::new(builder.finish()))
+    }
+}
+
+fn write_wkb_start_point(
```
Comment (paleolimbot) [inline rust/sedona-functions/src/st_start_point.rs]: Optional nit:

```suggestion
fn write_wkb_coord(

```

File: rust/sedona-functions/src/st_start_point.rs [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+    #[rstest]
+    fn udf(#[values(WKB_GEOMETRY, WKB_VIEW_GEOMETRY)] sedona_type: SedonaType) {
+        let tester_start_point =
+            ScalarUdfTester::new(st_start_point_udf().into(), vec![sedona_type.clone()]);
+        let tester_end_point =
+            ScalarUdfTester::new(st_end_point_udf().into(), vec![sedona_type.clone()]);
+
+        let input = create_array(
+            &[
+                Some("LINESTRING (1 2, 3 4, 5 6)"),
```
Comment (paleolimbot) [inline rust/sedona-functions/src/st_start_point.rs:224]: For full test coverage of your implementation I think we would need to test empty geometries here as well.

**Review by paleolimbot — APPROVED** [review]
It would be good to take care of the documentation and test suggestions here but I this is a great implementation and we can also handle that in a follow-on ticket if needed. Thank you!

**Comment by yutannihilation** [comment]
Thanks for the great review! Adding tests about empty geometry actually helped me to find a bug in my implementation of `ST_EndPoint` :) I think I addressed all the comments now.

**Review by paleolimbot — APPROVED** [review]
Thank you!
